### PR TITLE
fix(wget_agent): Prevent possible memory corruption

### LINKED ISSUE #1424: Warnings in building wget_agent on GCC8
### Logs

```lisp
wget_agent.c: In function ‘GetURL’:
wget_agent.c:319:37: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2041 [-Wformat-overflow=]
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
                                     ^~   ~~~~~~~~~~~~~~~~~
wget_agent.c:319:3: note: ‘sprintf’ output between 8 and 2055 bytes into a destination of size 2048
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:378:29: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2046 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
                             ^~
         proxy, WgetArgs,TempFileDirectory,TaintedURL,GlobalParam, no_proxy);
         ~~~~~                
wget_agent.c:378:5: note: ‘snprintf’ output between 36 and 10330 bytes into a destination of size 2047
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         proxy, WgetArgs,TempFileDirectory,TaintedURL,GlobalParam, no_proxy);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:383:29: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2046 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
                             ^~
         proxy, WgetArgs, TempFileDir, TaintedURL, GlobalParam, no_proxy);
         ~~~~~                
wget_agent.c:383:5: note: ‘snprintf’ output 36 or more bytes (assuming 8283) into a destination of size 2047
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s -P '%s' '%s' %s %s 2>&1",
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         proxy, WgetArgs, TempFileDir, TaintedURL, GlobalParam, no_proxy);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:388:29: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2046 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s '%s' %s %s 2>&1",
                             ^~
         proxy, WgetArgs,TaintedURL, GlobalParam, no_proxy);
         ~~~~~                
wget_agent.c:388:5: note: ‘snprintf’ output between 30 and 8277 bytes into a destination of size 2047
     snprintf(CMD,MAXCMD-1," %s /usr/bin/wget -q %s '%s' %s %s 2>&1",
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         proxy, WgetArgs,TaintedURL, GlobalParam, no_proxy);
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:424:41: warning: ‘/’ directive output may be truncated writing 1 byte into a region of size between 0 and 2047 [-Wformat-truncation=]
     snprintf(TempFilePath, MAXCMD-1, "%s/%s", TempFileDirectory, TaintedURL + Position);
                                         ^
wget_agent.c:424:5: note: ‘snprintf’ output 2 or more bytes (assuming 2049) into a destination of size 2047
     snprintf(TempFilePath, MAXCMD-1, "%s/%s", TempFileDirectory, TaintedURL + Position);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:454:37: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2041 [-Wformat-truncation=]
       snprintf(CMD,MAXCMD-1, "find '%s' -type f -exec mv {} %s \\; > /dev/null 2>&1", TempFileDirectory, TempFile);
                                     ^~                                                ~~~~~~~~~~~~~~~~~
wget_agent.c:454:7: note: ‘snprintf’ output 49 or more bytes (assuming 2096) into a destination of size 2047
       snprintf(CMD,MAXCMD-1, "find '%s' -type f -exec mv {} %s \\; > /dev/null 2>&1", TempFileDirectory, TempFile);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:439:37: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2043 [-Wformat-truncation=]
         snprintf(CMD,MAXCMD-1, "mv '%s' '%s' 2>&1", TempFilePath, TempFile);
                                     ^~              ~~~~~~~~~~~~
wget_agent.c:439:9: note: ‘snprintf’ output 14 or more bytes (assuming 2061) into a destination of size 2047
         snprintf(CMD,MAXCMD-1, "mv '%s' '%s' 2>&1", TempFilePath, TempFile);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:431:39: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2041 [-Wformat-truncation=]
         snprintf(CMD,MAXCMD-1, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", TempFilePath);
                                       ^~                                                                   ~~~~~~~~~~~~
wget_agent.c:431:9: note: ‘snprintf’ output between 70 and 2117 bytes into a destination of size 2047
         snprintf(CMD,MAXCMD-1, "find '%s' -mindepth 1 -type d -empty -exec rmdir {} \\; > /dev/null 2>&1", TempFilePath);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:435:51: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2031 [-Wformat-truncation=]
         snprintf(CMD,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, TempFilePath);
                                                   ^~                             ~~~~~~~~~~~~
wget_agent.c:435:9: note: ‘snprintf’ output 33 or more bytes (assuming 2080) into a destination of size 2047
         snprintf(CMD,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", TempFile, TempFilePath);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘SetEnv’:
wget_agent.c:638:3: warning: ‘strncpy’ specified bound 2048 equals destination size [-Wstringop-truncation]
   strncpy(GlobalParam, S, sizeof(GlobalParam)); // get the parameters, kind of " -A rpm -R fosso -l 1* "
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘Archivefs’:
wget_agent.c:714:33: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2043 [-Wformat-truncation=]
     snprintf(CMD,MAXCMD-1, "tar %s -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalParam, TempFile, Path);
                                 ^~                                    ~~~~~~~~~~~
wget_agent.c:714:5: note: ‘snprintf’ output 34 or more bytes (assuming 2081) into a destination of size 2047
     snprintf(CMD,MAXCMD-1, "tar %s -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalParam, TempFile, Path);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘replace_url_with_auth’:
wget_agent.c:880:44: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size between 2037 and 2046 [-Wformat-truncation=]
     snprintf(GlobalURL, FILEPATH, "%s%s:%s@%s", http, username, password, URI);
                                            ^~                             ~~~
wget_agent.c:880:5: note: ‘snprintf’ output 3 or more bytes (assuming 2059) into a destination of size 2048
     snprintf(GlobalURL, FILEPATH, "%s%s:%s@%s", http, username, password, URI);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:868:5: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
     strncpy(http, GlobalURL, strlen(GlobalURL) - strlen(URI));
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:868:30: note: length computed here
     strncpy(http, GlobalURL, strlen(GlobalURL) - strlen(URI));
                              ^~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘GetVersionControlCommand’:
wget_agent.c:950:33: warning: ‘.dir’ directive writing 4 bytes into a region of size between 1 and 2048 [-Wformat-overflow=]
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
                                 ^~~~
wget_agent.c:950:3: note: ‘sprintf’ output between 5 and 2052 bytes into a destination of size 2048
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:967:72: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2003 [-Wformat-overflow=]
       sprintf(command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
                                                                        ^~                                            ~~~~~~~~~
wget_agent.c:967:7: note: ‘sprintf’ output 65 or more bytes (assuming 8253) into a destination of size 2048
       sprintf(command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:969:35: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2038 [-Wformat-overflow=]
       sprintf(command, "git clone %s %s %s >/dev/null 2>&1 && rm -rf %s/.git", GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
                                   ^~                                           ~~~~~~~~~
wget_agent.c:969:7: note: ‘sprintf’ output between 45 and 8233 bytes into a destination of size 2048
       sprintf(command, "git clone %s %s %s >/dev/null 2>&1 && rm -rf %s/.git", GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:959:136: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 1941 [-Wformat-overflow=]
       sprintf(command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, TempFileDirectory);
                                                                                                                                        ^~                                                                         ~~~~~~~~~
wget_agent.c:959:7: note: ‘sprintf’ output 142 or more bytes (assuming 6283) into a destination of size 2048
       sprintf(command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:961:36: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2037 [-Wformat-overflow=]
       sprintf(command, "svn export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalURL, GlobalParam, TempFileDirectory);
                                    ^~                                         ~~~~~~~~~
wget_agent.c:961:7: note: ‘sprintf’ output between 46 and 6187 bytes into a destination of size 2048
       sprintf(command, "svn export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalURL, GlobalParam, TempFileDirectory);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c: In function ‘GetVersionControl’:
wget_agent.c:508:33: warning: ‘.dir’ directive writing 4 bytes into a region of size between 1 and 2048 [-Wformat-overflow=]
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
                                 ^~~~
wget_agent.c:508:3: note: ‘sprintf’ output between 5 and 2052 bytes into a destination of size 2048
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:509:37: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2041 [-Wformat-overflow=]
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
                                     ^~   ~~~~~~~~~~~~~~~~~
wget_agent.c:509:3: note: ‘sprintf’ output between 8 and 2055 bytes into a destination of size 2048
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:508:33: warning: ‘.dir’ directive writing 4 bytes into a region of size between 1 and 2048 [-Wformat-overflow=]
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
                                 ^~~~
wget_agent.c:508:3: note: ‘sprintf’ output between 5 and 2052 bytes into a destination of size 2048
   sprintf(TempFileDirectory, "%s.dir", GlobalTempFile);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:509:37: warning: ‘%s’ directive writing up to 2047 bytes into a region of size 2041 [-Wformat-overflow=]
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
                                     ^~   ~~~~~~~~~~~~~~~~~
wget_agent.c:509:3: note: ‘sprintf’ output between 8 and 2055 bytes into a destination of size 2048
   sprintf(DeleteTempDirCmd, "rm -rf %s", TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wget_agent.c:535:41: warning: ‘%s’ directive output may be truncated writing up to 2047 bytes into a region of size 2037 [-Wformat-truncation=]
   snprintf(command,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, TempFileDirectory);
                                         ^~                           ~~~~~~~~~~~~~~
wget_agent.c:535:3: note: ‘snprintf’ output between 33 and 4127 bytes into a destination of size 2047
   snprintf(command,MAXCMD-1, "tar -cf  '%s' -C '%s' ./ 1>/dev/null", GlobalTempFile, TempFileDirectory);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main.c: In function ‘main’:
main.c:211:5: warning: ‘strncpy’ specified bound 2048 equals destination size [-Wstringop-truncation]
     strncpy(GlobalURL,argv[arg],sizeof(GlobalURL));
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by andi8086**
Fix wget_agent's obviously possible memory corruption

Closes #1424 

## Description

A lot of snprintf/printfs are replaced by asprintf, which allocates memory dynamically and sufficiently.

### Changes

* Unneeded memset is removed
* Unneeded malloc is removed
* Some pointer checks are introduced
* Some memory leaks are corrected
* New error code for out of memory error
* Some simplifications in program flow
* One minor change in wget_agent.c where rc was used in an error message instead of rc_system
* Lots of new asprintf / free (this most probably needs review!!!)

### CODE REVIEW

File: src/wget_agent/agent/wget_agent.h [hunk 1]
```diff
@@ -41,26 +41,34 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define stat64(x,y) stat(x,y)
 #define systemError(line, error, cmd)  LOG_FATAL("[%s:%d] Error exit: %d, CMD:%s\n", __FILE__, line, WEXITSTATUS(error), cmd);
 
+#define MAXCMD  2048
```
Comment (maxhbr) [inline src/wget_agent/agent/wget_agent.h]: maybe find a better name for that variable

File: src/wget_agent/agent/wget_agent.c [hunk 2]
```diff
@@ -918,24 +1096,22 @@ void MaskPassword()
...
     // Create new parameters with masked password
     sprintf(newParam, " --username %s --password ****", username);
-    memset(GlobalParam, '\0', MAXCMD);
-    strcpy(GlobalParam, newParam);
-  }
-  // GIT
-  else {
+    memset(g.param, '\0', MAXCMD);
+    strcpy(g.param, newParam);
+  } else { // GIT
```
Comment (maxhbr) [inline src/wget_agent/agent/wget_agent.c]:
```suggestion
  } 
  else
  { // GIT
```

**Comment by andi8086** [comment]
Error in travis-ci has nothing to do with my change.

**Comment by mcjaeger** [comment]
"only" PHP copy past detector is failing for a different reason.

**Comment by GMishx** [comment]
@andi8086, can you please incorporate the changes suggested by @maxhbr so we can merge this PR?

File: src/wget_agent/agent/wget_agent.c [hunk 3]
```diff
@@ -948,30 +1149,60 @@ void MaskPassword()
...
+    }
   }
   else if (0 == strcmp(GlobalType, Type[1]))
   {
     replace_url_with_auth();
     if (GlobalProxy[0] && GlobalProxy[0][0])
-      sprintf(command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, TempFileDirectory, TempFileDirectory);
+    {
+      res = asprintf(&command, "git config --global http.proxy %s && git clone %s %s %s  && rm -rf %s/.git", GlobalProxy[0], GlobalURL, GlobalParam, tmpfile_dir, tmpfile_dir);
+      if (res == -1)
```
[the remark this answers is not in the document]
Comment (GMishx) [reply src/wget_agent/agent/wget_agent.c]: Solved in #1594

File: src/wget_agent/agent/wget_agent.c [hunk 4]
```diff
@@ -948,30 +1149,60 @@ void MaskPassword()
...
+  }
 
   if(withPassword < 0) MaskPassword();
   if (0 == strcmp(GlobalType, Type[0]))
   {
     if (GlobalProxy[0] && GlobalProxy[0][0])
-      sprintf(command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, TempFileDirectory);
+    {
+      res = asprintf(&command, "svn --config-option servers:global:http-proxy-host=%s --config-option servers:global:http-proxy-port=%s export %s %s %s --no-auth-cache >/dev/null 2>&1", GlobalProxy[4], GlobalProxy[5], GlobalURL, GlobalParam, tmpfile_dir);
+      if (res == -1)
```
[the remark this answers is not in the document]
Comment (GMishx) [reply src/wget_agent/agent/wget_agent.c]: Solved in #1594

**Pull request merged**